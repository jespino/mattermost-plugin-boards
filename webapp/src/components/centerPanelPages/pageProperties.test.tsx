// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import '@testing-library/jest-dom'
import {act, render} from '@testing-library/react'

import React from 'react'
import {Provider as ReduxProvider} from 'react-redux'

import {TestBlockFactory} from '../../test/testBlockFactory'
import {mockDOM, mockStateStore, wrapIntl} from '../../testUtils'

import PageProperties from './pageProperties'

beforeAll(() => {
    mockDOM()
})

describe('components/pageProperties', () => {
    const board = TestBlockFactory.createBoard()
    board.id = 'test-id'
    board.cardProperties[0].type = 'select'
    board.cardProperties[0].options = [{id: 'test', value: 'test', color: 'propColorBrown'}]
    board.cardProperties[1].type = 'text'
    board.cardProperties[2].type = 'text'
    board.cardProperties[3].type = 'text'
    const state = {
        users: {
            boardUsers: {
                1: {username: 'abc'},
                2: {username: 'd'},
                3: {username: 'e'},
                4: {username: 'f'},
                5: {username: 'g'},
            },
        },
        teams: {
            current: {id: 'team-id'},
        },
        boards: {
            current: board.id,
            boards: {
                [board.id]: board,
            },
            myBoardMemberships: {
                [board.id]: {userId: 'user_id_1', schemeAdmin: true},
            },
        },
        clientConfig: {
            value: {},
        },
    }
    const store = mockStateStore([], state)

    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('should match snapshot without value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperties
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot with value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {
            [board.cardProperties[0].id]: board.cardProperties[0].options[0].id,
            [board.cardProperties[1].id]: 'Test value 1',
            [board.cardProperties[2].id]: 'Test value 2',
            [board.cardProperties[3].id]: 'Test value 3',
        }
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperties
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot readonly without value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperties
                        page={page}
                        board={board}
                        readonly={true}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot readonly with value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {
            [board.cardProperties[0].id]: board.cardProperties[0].options[0].id,
            [board.cardProperties[1].id]: 'Test value 1',
            [board.cardProperties[2].id]: 'Test value 2',
            [board.cardProperties[3].id]: 'Test value 3',
        }
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperties
                        page={page}
                        board={board}
                        readonly={true}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot without edit board permissions', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {
            [board.cardProperties[0].id]: board.cardProperties[0].options[0].id,
            [board.cardProperties[1].id]: 'Test value 1',
            [board.cardProperties[2].id]: 'Test value 2',
            [board.cardProperties[3].id]: 'Test value 3',
        }
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperties
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={false}
                        canEditBoardCards={true}
                        newTemplateId={''}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot without edit card permissions', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {
            [board.cardProperties[0].id]: board.cardProperties[0].options[0].id,
            [board.cardProperties[1].id]: 'Test value 1',
            [board.cardProperties[2].id]: 'Test value 2',
            [board.cardProperties[3].id]: 'Test value 3',
        }
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperties
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={true}
                        canEditBoardCards={false}
                        newTemplateId={''}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })
})
