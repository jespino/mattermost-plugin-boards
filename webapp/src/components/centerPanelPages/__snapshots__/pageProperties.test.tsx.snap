// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pageProperties should match snapshot readonly with value 1`] = `
<div>
  <div
    class="PageProperties"
  >
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Status
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label propColorBrown "
        >
          <span
            class="Label-text"
          >
            test
          </span>
        </span>
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 1
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      >
        Test value 1
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 2
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      >
        Test value 2
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 3
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      >
        Test value 3
      </div>
    </div>
  </div>
</div>
`;

exports[`components/pageProperties should match snapshot readonly without value 1`] = `
<div>
  <div
    class="PageProperties"
  >
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Status
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label empty "
        >
          <span
            class="Label-text"
          >
            Empty
          </span>
        </span>
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 1
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      >
        value1
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 2
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      />
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 3
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      />
    </div>
  </div>
</div>
`;

exports[`components/pageProperties should match snapshot with value 1`] = `
<div>
  <div
    class="PageProperties"
  >
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Status
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label propColorBrown "
        >
          <span
            class="Label-text"
          >
            test
          </span>
        </span>
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 1
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title="Test value 1"
        value="Test value 1"
      />
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 2
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title="Test value 2"
        value="Test value 2"
      />
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 3
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title="Test value 3"
        value="Test value 3"
      />
    </div>
  </div>
</div>
`;

exports[`components/pageProperties should match snapshot without edit board permissions 1`] = `
<div>
  <div
    class="PageProperties"
  >
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Status
      </div>
      <div
        class="octo-propertyvalue"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label propColorBrown "
        >
          <span
            class="Label-text"
          >
            test
          </span>
        </span>
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 1
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title="Test value 1"
        value="Test value 1"
      />
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 2
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title="Test value 2"
        value="Test value 2"
      />
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        class="octo-propertyname octo-propertyname--readonly"
      >
        Property 3
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title="Test value 3"
        value="Test value 3"
      />
    </div>
  </div>
</div>
`;

exports[`components/pageProperties should match snapshot without edit card permissions 1`] = `
<div>
  <div
    class="PageProperties"
  >
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Status
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label propColorBrown "
        >
          <span
            class="Label-text"
          >
            test
          </span>
        </span>
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 1
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      >
        Test value 1
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 2
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      >
        Test value 2
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 3
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue octo-propertyvalue--readonly"
      >
        Test value 3
      </div>
    </div>
  </div>
</div>
`;

exports[`components/pageProperties should match snapshot without value 1`] = `
<div>
  <div
    class="PageProperties"
  >
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Status
            </span>
          </button>
        </div>
      </div>
      <div
        class="octo-propertyvalue"
        data-testid="select-non-editable"
        tabindex="0"
      >
        <span
          class="Label empty "
        >
          <span
            class="Label-text"
          >
            Empty
          </span>
        </span>
      </div>
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 1
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title="value1"
        value="value1"
      />
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 2
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title=""
        value=""
      />
    </div>
    <div
      class="PageProperty octo-propertyrow"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper"
        role="button"
      >
        <div
          class="octo-propertyname"
        >
          <button
            class="Button"
            type="button"
          >
            <span>
              Property 3
            </span>
          </button>
        </div>
      </div>
      <input
        class="Editable octo-propertyvalue"
        placeholder="Empty"
        spellcheck="true"
        style="width: 100%;"
        title=""
        value=""
      />
    </div>
  </div>
</div>
`;
